src/Spectre.libWavelet/WaveletDecomposerRef.h

```diff
@@ -0,0 +1,58 @@
+/*
+* WaveletDecomposerRef.h
+* Reference implementation of wavelet decomposer.
+*
+Copyright 2018 Michal Gallus
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+#pragma once
+#include "Convolution.h"
+#include "DataTypes.h"
+
+namespace spectre::algorithm::wavelet
+{
+constexpr unsigned int WAVELET_LEVELS = 10;
+constexpr unsigned int WAVELET_BASIS = 4;
+
+struct WaveletCoefficients
+{
+    CoefficientsPerLevel data[WAVELET_LEVELS + 1]; // + 1 for lowest frequency
+};
+
+/// <summary>
+/// Decomposes the signal into set of Daubechies coefficients.
+/// </summary>
+class WaveletDecomposerRef
+{
+public:
+    /// <summary>
+    /// Initializes a new instance of the <see cref="WaveletDecomposer"/> class.
+    /// </summary>
+    explicit WaveletDecomposerRef();
+    /// <summary>
+    /// Decomposes the signal into wavelet coefficients.
+    /// </summary>
+    /// <param name="signal">Signal to decompose.</param>
+    /// <returns>Set of Daubechies wavelet coefficients.</returns>
+    WaveletCoefficients Decompose(Signal&& signal) const;
+
+private:
+    inline void WaveletDecomposerRef::ApplyFilters(
```

Comment:
Hmmm... I see a little bit different convention as compared to the one we were using until now (i.e. check `libGenetic`). In general, we try to consider a function to be *function* in mathematical sense, mapping between input and output. I know, we may consider `coefficients` and `lowFrequencyCoefficients` to be outputs, but its calling convention is different than we have in most of the code.

I'll accept any way we choose - shame, you weren't available when `libGenetic` was built. We could agree on a single version back then. Especially, that I see, that your style is much more popular in general. Either way, I think consistency (at least to ourselves) is one of the crucial points we have to keep.

Reply:
Oh well, I didn't know that. I made it in such a way to avoid data copies, and keep the last bits of cache locality. All there is to it, is that it's just more efficient. I do understand that the it may feel more natural to have it the math-function-resembling manner, but after all, is it a necessity that algorithms that do not interleave need to have the same convention?

Reply:
Ok, fine - but we can always `std::move` the data, can't we? :)

Reply:
...Which obfuscates the code, yes we do. And which mostly works on heap-allocated data, and sometimes (depending on scope) in stack.

Reply:
As we spoke outside, I'll accept this since it affects 2 `private inline` functions and API does not get inconsistent.